copy/copy.go

```diff
@@ -1322,32 +1336,35 @@ func (ic *imageCopier) copyLayer(ctx context.Context, srcInfo types.BlobInfo, to
 	}
 	defer srcStream.Close()
 
-	bar := ic.c.createProgressBar(pool, false, srcInfo, "blob", "done")
+	return func() (types.BlobInfo, digest.Digest, error) { // A scope for defer
```

Comment:
This is just reinforcing the general pattern:

```go
func() {
   bar := ic.c.createProgressBar() 
   defer bar.Abort(false)
   // use bar
   // definite defer point, not “sometime in the future”
}
```

In this case, where we return just after we stop using the bar, it indeed makes no difference to the actual sequence of calls.

Reply:
I am OK to keep it as is if you find it more readable, but it comes at a (very) low cost.

Reply:
I don’t find the `func()` boilerplate more readable at all, but we’d need a language with generics and trailing closure syntax to improve on that.

My primary concern is reliability, and maintainability vs. future copy&pasted code; a small readability hit to get reliability is acceptable.

WRT performance, in this code that is limited by network bandwidth and latency, I don’t worry about a few once-per-layer function calls one bit (and a compiler could inline them, in principle).

---

Actually I would find it quite persuasive to argue that many of the functions in this file grew too long, and that progress bar scope is also a good candidate for splitting into a separate function. But the copy code has no tests, so I’d prefer to work on _that_ first (and I did start writing some tests, though I’m very far from done) before large-scale aesthetic reorganization.

Reply:
I concur on the size issues of functions in this file. Starting with tests seems like a very wise decision. I think it will also feel like a open-heart surgery after but the tests would be a huge help.